Screenshot functionality

Mambouna opened the pull request:
Creates new functionality in `storage.py` to take screenshots.
- Also rearranges some former `Storage()` functions to make working with multiple classes that need to work with files more ergonomic.
- Fixed a bug in `storage.py` that would prevent saving manually from a game script.
- Initialization of the `screenshot` instance under `storage.py` added to `runner.py`.
- Added the function `screen.screenshot()` for manually controlled screenshots.
- Added F12 as a default keybinding in `game.py` to take screenshots.

Fixes #290

Mambouna committed:
FIX #290 - Screenshot functionality

Creates new functionality in `storage.py` to take screenshots.
- Also rearranges some former `Storage()` functions to make working with multiple classes that need to work with files more ergonomic.
- Fixed a bug in `storage.py` that would prevent saving manually from a
  game script.
- Initialization of the `screenshot` instance under `storage.py` added
  to `runner.py`.
- Added the function `screen.screenshot()` for manually controlled
  screenshots.
- Added F12 as a default keybinding in `game.py` to take screenshots.

Mambouna committed:
FIX #290 - flake8 linting adjustments

Mambouna committed:
FIX #290 - Updated documentation with screenshot function

Mambouna commented:
Locally flake8 doesn't have any problems anymore and the documentation is updated. Want to give it another go?

lordmauve commented:
There's a lot here, but I think it would be better if screenshots were saved to a standard "Pictures" directory for each platform, and I am hesitant about exposing screenshot() as an API and also giving it a default key binding. I think F12 should be sufficient.

I don't think this needs to be wired through the Storage classes, but what were you aiming to achieve with that?

Mambouna commented:
We can also save to pictures directory of course. I used storage since it already had functionality useful for checking validity of directories and setting up proper paths per system. Since screenshots are files to be stored locally, I also thought it idiomatic to use the existing infrastructure and from my experience, games use different approaches. Some save screenshots to system picture folders, other do so in their own user data directories. Since there was no clear "standard" case, I went with this one.

**EDIT:** Also if there's more stuff in the future that has to deal with files, having reworked storage.py means that it's easier to integrate coming stuff too. But of course this would only be true if `storage.py` becomes sort of the central file where all file system stuff is done.

The systems can be separated of course, but then we'd likely still need to import or rewrite those utility functions as far as I can see. If you would prefer that though, I'm not opposed to it.

The user function being exposed was simply done to not restrict options. This way, if a user introduces a system of custom keybindings for example or just wants to give a different key as an easier to reach option where maybe you want to take a lot of screenshots, there's something provided already. I couldn't see any harm in introducing it, but again if you feel differently, I'm not inherently against removing it.

lordmauve commented:
I think we do need to save to "user library" directories because "~/.config/pgzero" is not a place anyone will find a screenshot. We also need to print the path it was saved to.

Mambouna committed:
FIX #290 - Screenshots now saved in user folder

Based on review by lordmauve, screenshots are now saved in the user home
directory under the subfolder `Pictures`. Because this negates some of
the former justification to make `storage.py` functionality more
generalized, these changes were reverted unless still required with this
change.

Mambouna commented:
> I think we do need to save to "user library" directories because "~/.config/pgzero" is not a place anyone will find a screenshot. We also need to print the path it was saved to.

You're correct. I've made the changes. Screenshots now saved in user pictures directory, a message is printed stating the file name and directory and I've updated the documentation to also state the default keybinding F12.

lordmauve commented on src/pgzero/storage.py:
```diff
@@ -189,3 +218,39 @@ def _get_json_error_keys(cls, obj, json_path='storage'):
+
+        # Creates the filename, made up of the script name and a timestamp.
+        timestamp = datetime.now().strftime("-%y%m%d-%H%M%S")
+        filename = self._project_name + timestamp + ".jpg"
```
Let's save screenshots as `.png` by default. We shouldn't save in a lossy format by default; people can convert to `.jpg` themselves if they are Ok with lossy compression.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/storage.py:
```diff
@@ -189,3 +218,39 @@ def _get_json_error_keys(cls, obj, json_path='storage'):
+        _ensure_path(self._path)
+
+        # Creates the filename, made up of the script name and a timestamp.
+        timestamp = datetime.now().strftime("-%y%m%d-%H%M%S")
```
```suggestion
        timestamp = datetime.now().strftime("-%Y-%m-%d_%H:%M:%S")
```

This is ISO format and more readable for humans.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/storage.py:
```diff
@@ -189,3 +218,39 @@ def _get_json_error_keys(cls, obj, json_path='storage'):
+    screenshots = Screenshots(project_name)
+
+
+class Screenshots:
```
I think all this screenshots code could move to `screen.py` and be totally decoupled from the storage system.

`storage` offers a specific capability very similar to the browser's `localStorage` object; it isn't a generic interface to file storage.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/storage.py:
```diff
@@ -23,17 +37,33 @@ def _get_platform_pgzero_path():
+
+    Under Windows, this is %USERPROFILE%\Pictures\pgzero. Under Linux/MacOS,
+    it's ~/Pictures/pgzero."""
+    if platform.system() == "Windows":
```
I don't know why we used `platform.system()` before but it isn't supported in WebAsssembly builds (it doesn't exist, raises AttributeError). We should use `sys.platform`.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/storage.py:
```diff
@@ -23,17 +37,33 @@ def _get_platform_pgzero_path():
+    if platform.system() == "Windows":
+        home = _get_windows_environ("USERPROFILE")
+        return os.path.join(home, "Pictures", "pgzero")
+    return os.path.expanduser(os.path.join("~", "Pictures", "pgzero"))
```
Let's not assume that platforms other than Windows are compatible with this scheme. I think we should list them explicitly. If there are platforms we haven't explicitly supported, we could warn and fall back to current working directory?

Mambouna replied:
Fixed. For now I've kept it to the same accepted platforms as I'm not familiar with how some others work. Including FreeBSD will probably be safe since it should be the same, others I don't know.

lordmauve commented on src/pgzero/storage.py:
```diff
@@ -23,17 +37,33 @@ def _get_platform_pgzero_path():
+    except IsADirectoryError:
+        pass
+    except FileExistsError:
+        pass
```
I think it's odd for these errors to pass silently. It might make sense for storage(?) but it makes less sense for screenshots. We should just let the error propagate.

Mambouna replied:
Fixed. The errors were passed because `os.makedirs()` throws an error by default if the directory already exists. Since this is just there to make sure it does and create folders if not, the errors should not create a crash.

I've left `storage.py` the same, but the screenshot functionality now uses the built-in parameter `exist_ok` on `os.makedirs()` to make this clearer.

lordmauve commented on src/pgzero/game.py:
```diff
@@ -256,6 +256,9 @@ def key_down(event):
                 sys.exit(0)
+            # Default key for screenshots is F12.
+            if event.key == pygame.K_F12:
+                pgzero.screen.screen_instance.screenshot()
```
If there were errors writing this screenshot we should catch the exception and print a traceback.

But only here - if the user calls `screen.screenshot()` they should receive the exception.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/storage.py:
```diff
@@ -189,3 +218,39 @@ def _get_json_error_keys(cls, obj, json_path='storage'):
+        # Save the screenshot.
+        pygame.image.save(surface, filepath)
+        print("Saved a screenshot named {} in the folder {}."
+              .format(filename, self._path))
```
If the user calls `screen.screenshot()` they should expect to receive the path instead of printing.

I think you should return the path from this function, return it in Screen.screenshot(), and do the print() in the F12 handler.

Mambouna replied:
Fixed.

Mambouna committed:
FIX #290 - Implemented review changes

- Screenshots saved as `.png`.
- Timestamp printed in ISO format.
- All functionality decoupled from `storage.py` and moved to `screen.py`.
- Using `sys.platform` instead of `platform.sys`.
- Only recognized OSes use user folder for screenshot, otherwise warns
  and uses CWD.
- Using `exist_ok` on `os.makedirs()` instead of passing errors
  manually.
- Using F12 to take screenshots catches errors and prints a traceback.
- User defined screenshots propagate the error instead.
- Only prints info about screenshots when using F12, otherwise returns
  filepath to user.

Mambouna commented:
Thanks for the review, I've implemented all your suggestions. I liked the idea of having a central spot for file system stuff at first, but you were right that this wasn't the time to start that by cobbling together individual parts.

lordmauve commented on src/pgzero/game.py:
```diff
@@ -256,6 +260,16 @@ def key_down(event):
+                          "Full path: {}".format(name, folder, path))
+                except Exception:
+                    print("ERROR while trying to take a screenshot with F12.")
+                    print(format_exc())
```
These need to go to stderr not stdout. `print_exc()` does this for you.

Mambouna replied:
Fixed.

lordmauve commented on doc/builtins.rst:
```diff
@@ -55,6 +55,16 @@ draw images to the screen ("blit" them).
+
+        Takes a screenshot of the entire game window and saves it.
+
+        On Windows, the save location is ``%USERPROFILE%\Pictures\pgzero``,
```
The `%USERPROFILE%` and `~` syntax is pretty cryptic as it requires much deeper knowledge about the platform. You could just say "`Pictures` in your home directory".

Mambouna replied:
Fixed.

lordmauve commented on doc/builtins.rst:
```diff
@@ -55,6 +55,16 @@ draw images to the screen ("blit" them).
+        on MacOS and Linux it is ``~/Pictures/pgzero``.
+
+        The default keybinding to take screenshots is F12 and doesn't
+        require the user to manually use this function in their code.
```
```suggestion
        You can press F12 to take a screenshot at any time, without
        needing to call this function.
```

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/game.py:
```diff
@@ -13,6 +13,10 @@
 
+# Traceback for screenshot errors.
+from traceback import format_exc
+import os.path
```
Imports are grouped in this file; imports from builtins should be in the first section.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/screen.py:
```diff
@@ -6,6 +6,12 @@
+import os
+from sys import platform
+from datetime import datetime
+import pygame.image
```
I'd prefer imports in three groups: stdlib, thirdparty, local.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/screen.py:
```diff
@@ -25,6 +31,63 @@ def make_color(arg):
+                           "screenshots. Please set the %USERPROFILE% "
+                           "environment variable.")
+        return os.path.join(home, "Pictures", "pgzero")
+    elif platform in ["linux", "linux2", "darwin"]:
```
I'm about 85% sure there is a peephole optimisation for this exact case in CPython, but lists must be allocated at runtime while tuples are interned in the code constants and reused, so for constant sequence literals you should use tuples:

```suggestion
    elif platform in ("linux", "linux2", "darwin"):
```

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/screen.py:
```diff
@@ -25,6 +31,63 @@ def make_color(arg):
+    else:
+        print("WARNING: Device platform {} not recognized, thus no user folder"
+              " found. Falling back to current directory to save screenshots."
+              .format(platform))
```
```suggestion
        print("WARNING: Device platform {} not recognized, thus no user folder"
              " found. Falling back to current directory to save screenshots."
              .format(platform),
              file=sys.stderr
        )
```

FWIW you can use f-strings and pathlib. Pygame Zero dates from before f-strings and pathlib but any versions of Python without them are long since end-of-life.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/screen.py:
```diff
@@ -25,6 +31,63 @@ def make_color(arg):
+
+        # Creates the filename, made up of the script name and a timestamp.
+        timestamp = datetime.now().strftime("-%Y-%m-%d_%H:%M:%S")
+        filename = self._project_name + timestamp + ".png"
```
```suggestion
        now = datetime.now()
        filename = f"{self._project_name}-{now:%Y-%m-%d_%H:%M:%S}.png"
```

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/screen.py:
```diff
@@ -25,6 +31,63 @@ def make_color(arg):
+def _initialize_screenshots(file_path):
+    # If the instance already exists, do nothing.
+    if "screenshots" in globals():
+        return
```
I don't think this check is needed. You can do this unconditionally, and doing so might be helpful in tests.

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/screen.py:
```diff
@@ -25,6 +31,63 @@ def make_color(arg):
+        file_path = os.path.abspath(file_path)
+    project_name, _ = os.path.splitext(os.path.basename(file_path))
+
+    global screenshots
```
As a point of style I prefer the `global` statement to be at the start of the function as if to declare "this function modifies the `screenshots` global".

Mambouna replied:
Fixed.

lordmauve commented on doc/builtins.rst:
```diff
@@ -55,6 +55,16 @@ draw images to the screen ("blit" them).
 
+    .. method:: screenshot()
+
+        Takes a screenshot of the entire game window and saves it.
```
```suggestion
        Takes a screenshot of the entire game window and saves it.
        
        Return the path of the file that was written, as a string.
```

Mambouna replied:
Fixed.

lordmauve commented on src/pgzero/game.py:
```diff
@@ -256,6 +260,16 @@ def key_down(event):
+                    path = pgzero.screen.screen_instance.screenshot()
+                    folder, name = os.path.split(path)
+                    print("Saved a screenshot named {} in the folder {}.\n"
+                          "Full path: {}".format(name, folder, path))
```
I think this could just be `f"Saved screenshot to {path}"`.

Mambouna replied:
Fixed.

lordmauve reviewed (changes requested):
The implementation is pretty close, just a few style nits.

I'd like to have a few simple tests of the path computation and writing screenshots (maybe not to my real `~/Pictures`, we can `unittest.mock.patch()` it to write to a TemporaryDirectory).

Mambouna committed:
FIX #290 - Partial fix of review requests

All suggestions from review have been incorporated, except for writing
the tests.

Mambouna committed:
FIX #290 - Fix of review requests

All suggestions from review have been incorporated.

Mambouna committed:
FIX #290 - Added tests for screenshots

Mambouna commented:
Okay, I tried adding relevant tests for the screenshot functionality. This was my best effort, as I don't really know what I'm doing as soon as tests go beyond simple asserts.

If this is unsatisfactory, I'm not sure I can improve upon it without very explicit instruction. While I would appreciate that, it might be the case that it would be less work overall if you edited the PR yourself then.

Mambouna commented:
@lordmauve Any further feedback?

lordmauve commented:
I think you got into a mess with merges and rebases as some of the commits on main were rebased onto your branch. I've interactively rebased to just keep your changes and force pushed. There's still a conflict, I'll try rebasing again.

lordmauve commented:
Ok. Tests are passing but incorrectly - they fail on Windows. I guess we traditionally didn't do a whole lot of platform-specific stuff, mostly just relying on Pygame to handle portability for us.

I think the root errors are

```
Traceback (most recent call last):
  File "C:\Users\lordm\dev\pgzero\test\test_screen.py", line 272, in test_take_screenshot
    self.screen.screenshot()
    ~~~~~~~~~~~~~~~~~~~~~~^^
  File "C:\Users\lordm\dev\pgzero\src\pgzero\screen.py", line 241, in screenshot
    return screenshots.take(self.surface)
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
  File "C:\Users\lordm\dev\pgzero\src\pgzero\screen.py", line 81, in take
    pygame.image.save(surface, filepath)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
pygame.error: Couldn't open C:\Users\lordm\AppData\Local\Temp\tmpj70qc5kqscreenshot_testdir\pgzero_screenshots\test_screen-2025-07-30_00:31:48.png
```

and

```
Traceback (most recent call last):
  File "C:\Users\lordm\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\unittest\case.py", line 58, in testPartExecutor
    yield
  File "C:\Users\lordm\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\unittest\case.py", line 651, in run
    self._callTestMethod(testMethod)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "C:\Users\lordm\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\unittest\case.py", line 606, in _callTestMethod
    if method() is not None:
       ~~~~~~^^
  File "C:\Users\lordm\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\unittest\mock.py", line 1426, in patched
    return func(*newargs, **newkeywargs)
  File "C:\Users\lordm\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\unittest\mock.py", line 1891, in _inner
    return f(*args, **kw)
  File "C:\Users\lordm\dev\pgzero\test\test_screen.py", line 256, in test_get_screenshot_path_linux
    self.assertEqual(result_path,
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
                     os.path.join("/home/user", "Pictures", "pgzero"))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lordm\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\unittest\case.py", line 907, in assertEqual
    assertion_func(first, second, msg=msg)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lordm\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\unittest\case.py", line 1273, in assertMultiLineEqual
    self.fail(self._formatMessage(msg, standardMsg))
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lordm\AppData\Roaming\uv\python\cpython-3.13.3-windows-x86_64-none\Lib\unittest\case.py", line 732, in fail
    raise self.failureException(msg)
AssertionError: 'C:\\Users\\lordm\\Pictures\\pgzero' != '/home/user\\Pictures\\pgzero'
- C:\Users\lordm\Pictures\pgzero
+ /home/user\Pictures\pgzero
```

Mambouna commented:
@lordmauve Thanks for the heads-up. I'll take a look at it probably the day after tomorrow. I don't have a windows system myself but the error looks like the actual platform is being misidentified so if that's the case, the fix shouldn't be too complex.

Mambouna committed:
Patched builtin modules for possible bug fix with tests failing per platform

Mambouna commented:
@lordmauve Okay so turns out `os.path` itself uses `sys.builtin_module_names` to check the platform. I tried adding a patch for that as well to the tests. This still runs fine on my end, could you try if this fixes the second error on Windows?

I've also tried to fix the permission problem, but that might be more tricky. Permissions in Windows for temporary files and directories from Python seem to be pretty wonky and whether a file can be opened seems to depend on whether it was opened at least once already. I tried reading a bit into the Pygame code but I don't really know C, so I'm not sure if `pygame.image.save()` opens, closes and opens the file. If yes, that seems to be the most likely culprit to me as it fits with the knows permission issues with `tempfile` from python.

The easiest fix in that case would be to simply have the test write to the actual user directory where all the permission weirdness isn't a problem, but that seems really icky for a testing suite. What do you think?

lordmauve commented:
> The easiest fix in that case would be to simply have the test write to the actual user directory where all the permission weirdness isn't a problem, but that seems really icky for a testing suite. What do you think?

I agree, that's icky. Might be Ok in GitHub Actions though.

lordmauve commented:
The tests are still failing on Windows :-(

Mambouna committed:
Patched os.path directly instead of patching builtin_module_names

Mambouna commented:
Hi, thanks for trying again, shame the last change didn't help...

Here's a last attempt that I don't think it actually very good practise but I think it shouldn't produce side effects. Instead of patching the values that _should_ inform which library python uses, I directly patch `os.path` itself with the given library.
Again, this runs fine for me and debugging suggests on Linux the switch of libraries occurs correctly (although that was also the case before). If this doesn't work, I'm not sure I have more ideas ...

Also the tests should still fail from the permission error, since I haven't found any solution for that yet. But if the other error is gone, that would be a good step forward.
